### ISSUE TITLE
out of memory using eval

### ISSUE THREAD

**Issue by fpvandoorn** [leanprover/lean3#460]
I don't know whether this is worth reporting, but when uncommenting the last line in the file below, I get the error (probably because Lean uses too much memory):

```
terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc
```

I was surprised that such a small example already used too much memory, but when unfolding everything, the term indeed becomes huge (for me, when replacing `C ×c D` by `C` everywhere, the term was accepted, but the output was indeed huge).

```
import algebra.category.constructions

open category eq category.ops functor prod

variables
{C D E : Precategory} (F : C ×c D ⇒ E)
(p : carrier (C ×c D))
(ff : carrier (C ×c D) → carrier E)

eval ap ff (refl p)
--eval ap (to_fun_ob F) (refl p)
```

Maybe we should limit the runtime of `eval` in some way?

**Comment by leodemoura** [comment]
I will add a command to limit the amount of time and memory.

BTW, we already have a command line option `-M` to limit the amount of memory.
Example:

```
lean -M 50 mem.hlean 
mem.hlean:11:0: error: excessive memory consumption detected at 'expression equality test' (potential solution: increase memory consumption thresold)
```

The argument for `-M` is the max amount of memory in megabytes. The option doesn't kill Lean, when the limit is reached, it interrupts the command, print the error message and continue.

The Emacs lean-mode allow us to provide extra command line options to Lean. I use `-M 100`.

**Comment by leodemoura** [comment]
@fpvandoorn I added a `set_option max_memory [value]`. The value is in megabytes.
I also added a default limit: 512Mb. I think the default is very generous. None of the library files consume more than 50Mb to be processed.
BTW, this limit does not include stack space and memory fragmentation.
We still have option `-M`. 

The example in your message terminates after 15 secs on my machine (using the default limit: 512Mb).
If we add `set_option max_memory 50` or use `-M 50`, then it terminates in 1.5secs.
